### PR TITLE
[typing] add internal DTypeLike specializations

### PR DESCRIPTION
**Pull request by jakevdp**

### CODE REVIEW

File: jax/_src/typing/dtypes.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  def dtype(self) -> DType: ...
+
+# DTypeLike is meant to annotate inputs to np.dtype that return
+# a valid JAX dtype. It's different than numpy.typing.DTypeLike
+# because JAX doesn't support objects or structured dtypes.
+# Unlike np.typing.DTypeLike, we exclude None, and instead require
+# explicit annotations when None is acceptable.
+# TODO(jakevdp): consider whether to add ExtendedDtype to the union.
+DTypeLike = Union[
+  str,            # like 'float32', 'int32'
```
Comment (superbobry) [inline jax/_src/typing/dtypes.py:37]: Would it be useful to make this a union of literal types iyo?

File: jax/_src/typing/dtypes.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+from __future__ import annotations
+
+from typing import Any, Literal, Protocol, Union
+import numpy as np
+import ml_dtypes
+
+DType = np.dtype
+
+# TODO(jakevdp, froystig): make ExtendedDType a protocol
+ExtendedDType = Any
```
Comment (superbobry) [inline jax/_src/typing/dtypes.py:24]: What does this type correspond to?

File: jax/_src/typing/dtypes.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+# DTypeLike is meant to annotate inputs to np.dtype that return
+# a valid JAX dtype. It's different than numpy.typing.DTypeLike
+# because JAX doesn't support objects or structured dtypes.
+# Unlike np.typing.DTypeLike, we exclude None, and instead require
+# explicit annotations when None is acceptable.
+# TODO(jakevdp): consider whether to add ExtendedDtype to the union.
+DTypeLike = Union[
+  str,            # like 'float32', 'int32'
+  type[Any],      # like np.float32, np.int32, float, int
+  np.dtype,       # like np.dtype('float32'), np.dtype('int32')
```
Comment (superbobry) [inline jax/_src/typing/dtypes.py:39]: Nit: `DType[Any]`?

File: jax/_src/typing/dtypes.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  @property
+  def dtype(self) -> DType: ...
+
+# DTypeLike is meant to annotate inputs to np.dtype that return
+# a valid JAX dtype. It's different than numpy.typing.DTypeLike
+# because JAX doesn't support objects or structured dtypes.
+# Unlike np.typing.DTypeLike, we exclude None, and instead require
+# explicit annotations when None is acceptable.
+# TODO(jakevdp): consider whether to add ExtendedDtype to the union.
+DTypeLike = Union[
```
Comment (superbobry) [inline jax/_src/typing/dtypes.py:36]: Why is it a bad idea to define `DTypeLike` as a union of the specialized `DTypeLike*` aliases below?

File: jax/_src/typing/dtypes.py (hunk 2 above)
Comment (jakevdp) [reply jax/_src/typing/dtypes.py:24]: `ExtendedDType` is used for `bint` and `prng_key` dtypes; it's discussed here: https://jax.readthedocs.io/en/latest/jep/9263-typed-keys.html#extended-dtypes

File: jax/_src/typing/dtypes.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+BFloat16Literals = Literal['bfloat16']
+Float16Literals = Literal['float16', 'f2']
+Float32Literals = Literal['float32', 'f4']
+Float64Literals = Literal['float', 'float64', 'f8']
+
+Complex64Literals = Literal['complex64', 'c8']
+Complex128Literals = Literal['complex', 'complex128', 'c16']
+
+# TODO(jakevdp): the use of things like type[float] and type[np.floating]
+#                below are not strictly correct: can we do better?
```
[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/typing/dtypes.py:64]: I think it's the opposite: we want this to include `Literal[int]` – i.e. integer instances are not allowed (`np.dtype(1)` fails) but the actual literal type `int` is allowed (`np.dtype(int)` succeeds).

[the remark this answers is not in the document]
Comment (superbobry) [reply jax/_src/typing/dtypes.py:64]: Ooops, yes, you are absolutely right! We want an invariant `type`.

File: jax/_src/typing/dtypes.py (hunk 1 above)
Comment (jakevdp) [reply jax/_src/typing/dtypes.py:37]: Possibly – this is a public API though, so making it more strictly defined will in general lead to downstream failures that will prevent us from landing this.

**Review by superbobry — APPROVED** [review]

File: jax/_src/typing/dtypes.py (hunk 3 above)
Comment (jakevdp) [reply jax/_src/typing/dtypes.py]: Done

File: jax/_src/typing/dtypes.py (hunk 4 above)
Comment (jakevdp) [reply jax/_src/typing/dtypes.py:37]: `DTypeLike` is used in public APIs, so if we change it there's a good chance the perturbation will break the builds of downstream packages. Plus the simplicity here is nice (even if there are invalid strings that it lets through).

**Comment by jakevdp** [comment]
I don't think we can merge this: the issue is that these specializations reject valid code if the value of the string input is unknown statically to the type checker.

The semantics I want are to allow any generic string in general, BUT if the string value is known statically, it must match one of the known values.

As far as I understand, Python's type annotation syntax cannot express that, so accepting any `str` is the best we can do.

**Pull request closed without merging**